### PR TITLE
feat: digest auth

### PR DESCRIPTION
**Pull request by Jakub-Vacek**
## Description

This PR adds digest authentication. Missing some tests and more importantly `user` and `password` inputs. Tested against real ClockPlus API.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] I have updated the documentation accordingly. If you are changing **code related to user secrets** you need to really make sure that [security documentation](SECURITY.md) is correct.
- [x] I have read the **CONTRIBUTION_GUIDE** document.
- [x] I haven't repeated the code. (DRY)
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: src/internal/interpreter/http/digest.ts [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+    if (match) return trim ? match[1].replace(/[\s"]/g, '') : match[1];
+
+    return undefined;
+  }
+
+  private static computeHash(
+    user: string,
+    realm: string,
+    password: string
+  ): string {
```
Comment (TheEdward162) [inline src/internal/interpreter/http/digest.ts:166]: If I understand it correctly this is supposed to be the default route for when no algorithm is specified? The name is vague and the algorithm is harcoded, so I'm not sure.

File: src/internal/interpreter/http/digest.ts [hunk 2]
```diff
@@ -0,0 +1,262 @@
...
+    private readonly cnonceSize: number = 32,
+    //407 can be also used when communicating thru proxy https://datatracker.ietf.org/doc/html/rfc2617#section-1.2
+    private readonly statusCode: number = 401,
+    //"Proxy-Authenticate" can be also used when communicating thru proxy https://datatracker.ietf.org/doc/html/rfc2617#section-1.2
+    private readonly header: string = 'www-authenticate',
+    private nc: number = 0
+  ) {
+    debug('Initialized DigestHelper');
+    debugSensitive(
+      `Initialized with: userame="${this.user}", password="${this.password}", cnonce size=${this.cnonceSize}, status code=${this.statusCode}, challenge header="${this.header}"`
```
Comment (jnv) [inline src/internal/interpreter/http/digest.ts:54]: Typo: userame

File: src/internal/interpreter/http/http.ts [hunk 3]
```diff
@@ -170,101 +234,91 @@ export class HttpClient {
       }
 
       if (configuration.type === SecurityType.APIKEY) {
-        applyApiKeyAuth(contextForSecurity, configuration);
+        const handler = new ApiKeyHandler()
+        handler.prepare(contextForSecurity, configuration)
```
Comment (lukas-valenta) [inline src/internal/interpreter/http/http.ts:238]: if you don't need to do stuff with the constructed unprepared handler, it'll be probably better to use static prepare for construction, i.e. `const handler = ApiKeyHandler.prepare(contextForSecurity, configuration)`

File: src/client/profile-provider.ts [hunk 4]
```diff
@@ -67,11 +67,12 @@ function profileAstId(ast: ProfileDocumentNode): string {
 const boundProfileProviderDebug = createDebug(
   'superface:bound-profile-provider'
 );
+export type AuthCache = { cache?: { digest?: string } };
```
Comment (lukas-valenta) [inline src/client/profile-provider.ts:70]: shouldn't there be some sort of cache key?

Comment (Jakub-Vacek) [reply src/client/profile-provider.ts:70]: This exists on boundProfileProvider so it is already scoped only on one provider. Not sure if something more is needed:
 - if user reuses boundProfileProvider with different auth. method this won't be used

File: src/internal/interpreter/http/http.ts (hunk 3 above)
Comment (Jakub-Vacek) [reply src/internal/interpreter/http/http.ts]: You mean the `ApiKeyHandler.prepare(contextForSecurity, configuration)` should change the `contextForSecurity` and return new `ApiKeyHandler` instance? It feels like two different things. But maybe the "change the `contextForSecurity`" part is hard to spot at first 🤔

Comment (lukas-valenta) [reply src/internal/interpreter/http/http.ts]: oh i didn't notice it changes the context. that's not not very good tbh :thinking:

Comment (lukas-valenta) [reply src/internal/interpreter/http/http.ts]: but it's okay now, with the configuration in constructor :shrug:

File: src/internal/interpreter/http/security/digest/digest.ts [hunk 5]
```diff
@@ -0,0 +1,296 @@
...
+  //Internal
+  private readonly nonceRaw: string = 'abcdef0123456789';
+  private readonly cnonceSize = 32;
+  private nc = 0;
+
+  constructor(
+    readonly configuration: DigestSecurityScheme & DigestSecurityValues
+  ) {
+    debug('Initialized DigestHelper');
+    this.statusCode = configuration.statusCode || 401;
```
Comment (TheEdward162) [inline src/internal/interpreter/http/security/digest/digest.ts:59]: Can we has `??` here?

File: src/internal/interpreter/http/security/digest/digest.ts [hunk 6]
```diff
@@ -0,0 +1,296 @@
...
+      qop: extractQop(header),
+      nonce,
+      cnonce: this.makeNonce(),
+    };
+  }
+
+  private makeNonce(): string {
+    let uid = '';
+    for (let i = 0; i < this.cnonceSize; ++i) {
+      uid += this.nonceRaw[Math.floor(Math.random() * this.nonceRaw.length)];
```
Comment (TheEdward162) [inline src/internal/interpreter/http/security/digest/digest.ts:203]: I'm not sure about the `Math.random` security implications here.

@lukas-valenta What do you think?

Comment (TheEdward162) [reply src/internal/interpreter/http/security/digest/digest.ts:203]: On the other hand, it is only used to compute cnonce.. 🤷

File: src/internal/interpreter/http/security/digest/digest.ts [hunk 7]
```diff
@@ -0,0 +1,296 @@
...
+      ]);
+    }
+  }
+
+  // when not specified
+  return undefined;
+}
+
+function extract(raw: string, field: string, trim = true): string | undefined {
+  //TODO: field should be case-insensitive
```
Comment (TheEdward162) [inline src/internal/interpreter/http/security/digest/digest.ts:273]: This todo no longer applies, right?

File: src/lib/fetch.ts [hunk 8]
```diff
@@ -65,9 +65,10 @@ export class CrossFetch implements FetchInstance, Interceptable {
     let body: unknown;
 
     if (
-      (headers['content-type'] &&
-        headers['content-type'].includes(JSON_CONTENT)) ||
-      parameters.headers?.['accept']?.includes(JSON_CONTENT)
+      headers['content-type'] &&
+      headers['content-type'].includes(JSON_CONTENT) //||
+      //This should be used when we don't have a response content-type
```
Comment (TheEdward162) [inline src/lib/fetch.ts:70]: Please make this a TODO for when security handlers get updated to make requests themselves, as we talked.

File: src/internal/interpreter/http/security/digest/digest.ts (hunk 6 above)
Comment (lukas-valenta) [reply src/internal/interpreter/http/security/digest/digest.ts:203]: i missed that. cnonce or no, let's use crypto

**Comment by Jakub-Vacek** [comment]
@lukas-valenta @TheEdward162 Technically, this is good to go - it does not depend on any rc/beta releases, it works and it is covered with tested. Can we merge it or do we want wait on events auth. architecture? 🤷🏻‍♂️

**Comment by jnv** [comment]
Please land it before it drifts away from main. 😅

**Comment by lukas-valenta** [comment]
@jnv too late for that :disappointed:

**Review by lukas-valenta — APPROVED** [review]

**Pull request merged**